src/gui/editors/PianoRoll.cpp

```diff
@@ -2830,8 +2858,96 @@ int PianoRoll::xCoordOfTick( int tick )
 
 void PianoRoll::paintEvent(QPaintEvent * pe )
 {
+	QStringList labelsList;
+	QStringList numbersList;
 	bool drawNoteNames = ConfigManager::inst()->value( "ui", "printnotelabels").toInt();
-
+	int customLabelsFile;
+	bool drawCustomLabels = 0;
+	if (m_labelsModel.value() == 0){
+		drawCustomLabels = 0;
+	} else if (m_labelsModel.value() == 1) {
+		drawCustomLabels = 1;
+		if (labelsFile1Correct){
+			customLabelsFile = 1;
+		} else if (labelsFile2Correct) {
+			customLabelsFile = 2;
+		} else {
+			customLabelsFile = 3;
+		}
+	} else if (m_labelsModel.value() == 2) {
+		drawCustomLabels = 1;
+		if (labelsFile1Correct){
+			if (labelsFile2Correct) {
+				customLabelsFile = 2;
+			} else {
+				customLabelsFile = 3;
+			}
+		} else {
+			customLabelsFile = 3;
+		}
+	} else if (m_labelsModel.value() == 3) {
+		drawCustomLabels = 1;
+		customLabelsFile = 3;
+	}
+	
+	QString labels1Path = ConfigManager::inst()->labels1Dir();
+	QString labels2Path = ConfigManager::inst()->labels2Dir();
+	QString labels3Path = ConfigManager::inst()->labels3Dir();
+	
+	if (drawCustomLabels){
+		QString labelsPath;
+		if(customLabelsFile == 1) labelsPath = labels1Path; 
+		if(customLabelsFile == 2) labelsPath = labels2Path; 
+		if(customLabelsFile == 3) labelsPath = labels3Path; 
+				
+		QFile labelsFile(labelsPath);
+		labelsFile.open(QIODevice::ReadOnly);
+		QTextStream stream(&labelsFile);
+		QString line = stream.readLine();
+		QRegExp num("\\d*");
+		while (!line.isNull()) {
```

Comment:
This happens *every* `paintEvent`, which will block on file I/O which can cause freezing of the UI in worse cases. These labels need to be read on LMMS load rather than here and painted from memory.

Reply:
Yes, I've noticed that. I know that is not good way, I'm thinking how to change that.